### PR TITLE
SubWindow: Increase to respect child's sizeHint

### PR DESCRIPTION
**Pull request by JohannesLorenz**
Before this commit, on creation, `SubWindow` gets resized to exactly the children's `sizeHint`. This makes the child too small, since the `SubWindow` already contains a title bar and borders.

With this commit, the `SubWindow` is calculated such that after rendering, the child window gets exactly the `size` that its `sizeHint` has previously suggested.

Most of LMMS widgets are not resizable, but the Lv2 help window is a good example to test this out. The help windows now in most cases contain enough space to fit the help text. In some cases, it still does not fit, though debug prints show that the `size` matches the `sizeHint`.

### CODE REVIEW

File: src/gui/SubWindow.cpp [hunk 1]
```diff
@@ -240,6 +241,27 @@ void SubWindow::setBorderColor( const QColor &c )
 
 
 
+
+int SubWindow::titleBarHeight() const
+{
+	QStyleOptionTitleBar so;
+	so.titleBarState = Qt::WindowActive; // kThemeStateActiv
+	so.titleBarFlags = Qt::Window;
+	return style()->pixelMetric(QStyle::PM_TitleBarHeight, &so, this);
```
Comment (JohannesLorenz) [inline src/gui/SubWindow.cpp:250]: I was not exactly sure whether to specify the 2nd and 3rd arguments (for both `pixelMetric` calls), though it just works.

File: src/gui/MainWindow.cpp [hunk 2]
```diff
@@ -570,13 +570,21 @@ void MainWindow::addSpacingToToolBar( int _size )
...
 SubWindow* MainWindow::addWindowedWidget(QWidget *w, Qt::WindowFlags windowFlags)
 {
 	// wrap the widget in our own *custom* window that patches some errors in QMdiSubWindow
 	auto win = new SubWindow(m_workspace->viewport(), windowFlags);
 	win->setAttribute(Qt::WA_DeleteOnClose);
 	win->setWidget(w);
-	if (w && w->sizeHint().isValid()) {win->resize(w->sizeHint());}
+	if (w && w->sizeHint().isValid()) {
+		std::size_t titleBarHeight = win->titleBarHeight();
+		std::size_t frameWidth = win->frameWidth();
```
Comment (Veratil) [inline src/gui/MainWindow.cpp:584]: You're using `std::size_t` here, but the return value is `int`. `QSize` takes `int` as constructor parameters. Should probably just make these `auto`.

Comment (JohannesLorenz) [reply src/gui/MainWindow.cpp]: No, `int` is one letter shorter. :grin:

Comment (Veratil) [reply src/gui/MainWindow.cpp]: Bad reason but I'm not going to argue.

**Comment by zonkmachine** [comment]
Nice!

> In some cases, it still does not fit, though debug prints show that the `size` matches the `sizeHint`.

Do you have an example of this? All the plugins help windows that I have resize well. Maybe the smallest ones look a bit tiny. If it's just one short sentence I think it could be written out in one line and with a bit of margin too.

File: src/gui/MainWindow.cpp (hunk 2 above)
Comment (JohannesLorenz) [reply src/gui/MainWindow.cpp]: Sorry, I think your argument makes sense. If Qt should ever switch from int to long, we are now prepared with `auto`.

**Comment by JohannesLorenz** [comment]
> Do you have an example of this?

The "infamous cellular automaton" (ssj71 plugins) still does not fit (the text is one word too big).
For JX10, the text is too small (one extra blank line is displayed). Though I just see that this effect has an extra blank line in the source code - so this one should be OK.

Possibly we should add some extra space for the strange case where the `QLabel::sizeHint` fails to calculate a useful size...

**Comment by zonkmachine** [comment]
Kars is also one line short. This is what the `getDescription()` line looks like. It looks like it misses a newline.
```
    const char* getDescription() const override
    {
        return "Simple karplus-strong plucked string synth.";
    }
```

Here is what it looks like with and without a newline added.

image

image

Maybe add a newline on the fly if the string doesn't end in one?

**Comment by JohannesLorenz** [comment]
> Maybe add a newline on the fly if the string doesn't end in one?

Funnily, this works, though still not 100% correctly. E.g. DX10 now shows a blank line, while the infamous cellular automaton does not (at least, if I combine this PR with #6957). However, it seems to be the best possible compromise - a blank line hurts less than a too small window which you need to resize? So I add a newline to #6957 .

**Comment by zonkmachine** [comment]
> Funnily, this works, though still not 100% correctly.

I think it works well enough. Looks good!

**Review by zonkmachine — APPROVED** [review]
Based on testing, approved.

**Comment by JohannesLorenz** [comment]
> Based on testing, approved.

@zonkmachine Thanks a lot. This PR is a bit critical, because it might enlarge *all* kinds of subwindows (mixer, instrument window, piano roll - everything). However, I assume many of those widgets will stay the same, since many are being resized manually. Did you check if other subwindows (not only Lv2) are enlarged in an unusual way?

**Comment by zonkmachine** [comment]
> Did you check if other subwindows (not only Lv2) are enlarged in an unusual way?

Well, yes, but I went and tested it some more. I haven't seen anything suspicious at all. Works just great.

**Comment by JohannesLorenz** [comment]
Testing the SubWindows should be a sufficient measure to verify this minor PR. Will merge on Sunday if no one complains.

**Pull request merged**